src/Npgsql/ConnectorPool.cs

```diff
@@ -413,125 +378,95 @@ internal void Release(NpgsqlConnector connector)
-                // may occur as we're putting our connector into the idle list. Decrement the busy
-                // count, while atomically make sure the waiting count isn't increased.
-                // Note that we also must update the state *before* putting the connector back in the idle list.
-                var newState = state;
-                newState.Idle++;
-                newState.Busy--;
-                CheckInvariants(newState);
-                if (Interlocked.CompareExchange(ref State.All, newState.All, state.All) != state.All)
-                {
-                    // Our attempt to decrement the busy count failed, either because a waiting attempt has been added
-                    // or busy has changed. Loop again and retry.
-                    continue;
-                }
-
-                // If we're here, we successfully applied the new state above and can put the connector back in the idle
-                // list (there were no pending open attempts).
 
-                connector.ReleaseTimestamp = DateTime.UtcNow;
-
-                // We start scanning for an empty slot in "random" places in the array, to avoid
-                // too much interlocked operations "contention" at the beginning.
-                var start = Thread.CurrentThread.ManagedThreadId % _max;
-
-                sw = new SpinWait();
-                while (true)
-                {
-                    for (var i = start; i < _idle.Length; i++)
-                    {
-                        if (Interlocked.CompareExchange(ref _idle[i], connector, null) == null)
-                        {
-                            Counters.NumberOfFreeConnections.Increment();
-                            return;
-                        }
-                    }
+                // If the open attempt timed out, the Task's state will be set to Canceled and our
+                // TrySetResult fails. Try again.
+                Debug.Assert(tcs.Task.IsCanceled);
+            }
 
-                    for (var i = 0; i < start; i++)
-                    {
-                        if (Interlocked.CompareExchange(ref _idle[i], connector, null) == null)
-                        {
-                            Counters.NumberOfFreeConnections.Increment();
-                            return;
-                        }
-                    }
-                    sw.SpinOnce();
-                }
+            // Scenario: pre-empted release
+            // Right here between our check for waiters and our signalling decrement for storing
+            // a connector there could have been a new waiter enqueueing, we compensate at the end.
+
+            // If we're here, we put the connector back in the idle list
+            // We increment Idle, any allocate that is racing us will not match Busy == _max
+            // and will not enqueue but try to get our connector.
+            Interlocked.Increment(ref State.Idle);
```

Comment:
@roji FYI thought about your concern for queue position fairness again. I had this suggestion up earlier as well.

If we're after this increment we could also check `_waiting` again to do an actual handover instead of an unblocking, essentially moving that code (line 414) up a bit. That would introduce a race similar to what's already happening anyway, Idle being incremented but having no connector actually be stored yet. 

Doing that has the downside that it would come with an extra decrement but it could very well be worth it.

Reply:
I see what you mean. To be honest, my point about loss of fairness was largely theoretical, I don't think the current code actually has any issues. What you're proposing would only occur during a race condition between Release and enqueuing into Waiting... It's very unlikely that the waiter would spend any significant time in the queue before losing its place, or that a significant number of other waiters would make into into the queue.

So I'd vote for staying with the code as-is - or do you think it's worth dealing with this?

Reply:
No it's why I left it like this :)